src/Build/Evaluation/Expander.cs

```diff
@@ -2956,6 +3146,128 @@ internal static IEnumerable<KeyValuePair<string, S>> AnyHaveMetadataValue(Expand
                         yield return new KeyValuePair<string, S>("false", null);
                     }
                 }
+
+                /// <summary>
+                /// Expands the metadata in the match provided into a string result.
+                /// The match is expected to be the content of a transform.
+                /// For example, representing "%(Filename.obj)" in the original expression "@(Compile->'%(Filename.obj)')".
+                /// </summary>
+                private static string GetMetadataValueFromMatch(
+                    MetadataMatch match,
+                    string itemSpec,
+                    IItem sourceOfMetadata,
+                    IElementLocation elementLocation,
+                    ref int curIndex)
+                {
+                    string value = null;
+                    try
+                    {
+                        if (FileUtilities.ItemSpecModifiers.IsDerivableItemSpecModifier(match.Name))
+                        {
+                            // If we're not a ProjectItem or ProjectItemInstance, then ProjectDirectory will be null.
+                            // In that case, we're safe to get the current directory as we'll be running on TaskItems which
+                            // only exist within a target where we can trust the current directory
+                            string directoryToUse = sourceOfMetadata.ProjectDirectory ?? Directory.GetCurrentDirectory();
```

Comment:
[unrelated to this PR] @AR-May we'll have to rethink this assumption for threading.